src/uwtools/cli.py

```diff
@@ -0,0 +1,605 @@
+        "--values-format",
+        choices=choices,
+        help="Values format",
+        required=required,
+        type=str,
+    )
+
+
+def add_arg_values_needed(group: Group) -> None:
+    group.add_argument(
+        "--values-needed",
+        action="store_true",
+        help="Print report of values needed to render template",
+    )
+
+
+def add_arg_verbose(group: Group) -> None:
+    group.add_argument(
+        "--verbose",
+        "-v",
+        action="store_true",
+        help="Print all logging messages",
+    )
+
+
+# pylint: enable=missing-function-docstring
+
+
+# Support
+
+
+def abort(msg: str) -> None:
+    """
+    Exit with an informative message and error status.
+
+    :param msg: The message to print.
+    """
+    print(msg, file=sys.stderr)
+    sys.exit(1)
+
+
+def add_subparser(subparsers: Subparsers, name: str, helpmsg: str) -> Parser:
+    """
+    Add a new subparser, with standard help formatting, to the given parser.
+
+    :param subparsers: The subparsers to add the new subparser to.
+    :param name: The name of the subparser.
+    :param helpmsg: The help message for the subparser.
+    :return: The new subparser.
+    """
+    return subparsers.add_parser(
+        name, add_help=False, help=helpmsg, formatter_class=formatter, description=helpmsg
+    )
+
+
+def add_subparsers(parser: Parser, dest: str) -> Subparsers:
+    """
+    Add subparsers to a parser.
+
+    :parm parser: The parser to add subparsers to.
+    :return: The new subparsers object.
+    """
+    return parser.add_subparsers(
+        dest=dest, metavar="MODE", required=True, title="Positional arguments"
+    )
+
+
+def basic_setup(parser: Parser) -> Group:
+    """
+    Create optional-arguments group and add help switch.
+
+    :param parser: The parser to add the optional group to.
+    """
+    optional = parser.add_argument_group("Optional arguments")
+    optional.add_argument("-h", "--help", action="help", help="Show help and exit")
+    return optional
+
+
+def check_args(args: Namespace) -> Namespace:
+    """
+    Validate basic argument correctness.
+
+    :param args: The parsed command-line arguments to check.
+    :return: The checked command-line arguments.
+    :raises: SystemExit if any checks failed.
+    """
+    try:
+        if args.quiet and args.verbose:
+            abort("Specify at most one of --quiet, --verbose")
+    except AttributeError:
+        pass
+    try:
+        if args.values_file and not args.values_format:
+            abort("Specify --values-format with --values-file")
+    except AttributeError:
+        pass
+    return args
+
+
+def dict_from_key_eq_val_strings(config_items: List[str]) -> Dict[str, str]:
+    """
+    Given a list of key=value strings, return a dictionary of key/value pairs.
+
+    :param config_items: Strings in the form key=value.
+    :return: A dictionary based on the input key=value strings.
+    """
+    return dict([arg.split("=") for arg in config_items])
```

Comment:
I'm still on the fence about all keys being a part of OUR language. But I think that for component configs, those keys also have a problem with variables having something like a `.` in them -- I'm thinking bash and Fortran variables, specifically.

Reply:
`.` might be ok, but note that what appears on the left of an `=` in a Fortran namelist doesn't just have to be a variable -- it can be an array reference, which might involve parens, colons, and commas; or an object-member reference, which involves `%`. I can't think of a case using `.` in a namelist LHS at the moment, but I haven't looked at what's permitted in namelists, especially in later Fortran standards, in a long time.

In general, though, if we don't keep UW DSL keys and target-language keys distinct, isn't any key in our UW YAML potentially a key in some other config language -- say, a key in another YAML language, or in an INI file? For sure a key in somebody else's YAML language could contain a `.`, and I guess that might be true of INI files, too. I mean, might we find that we need to create/edit a YAML or INI file that has a key with a dot in it? If so, that key could not appear as a key in UW YAML, as otherwise we wouldn't be able to distinguish between that key and a parent/child key pair separated by a dot.

It's just not sound to flatten a hierarchical language this way unless we can place certain constraints on the content. And if we don't control the key space, I don't think we can. I think we can only make an educated guess that the probability that we'll hit the buggy cases are low, and cross our fingers. Or solve a hierarchical problem using hierarchical tools.

Reply:
Could we take the `sed` approach and use a `.` by default, but then allow a user to specific a few different delimiters if that is not appropriate for their edge case?

Reply:
Sounds complicated/messy. It's unlikely that I will warm to this idea while I think I see robust alternatives. JSON strings are ugly but sound. I think the better solution is a YAML language where the keys are strictly DSL elements. We probably don't need an answer to this question soon, so have some time to think.